### ISSUE TITLE
Question - Cross-validation and Impact of overfitting

### ISSUE THREAD

**Issue by Jacques-Peeters** [uber/causalml#517]
👋 

**Is your feature request related to a problem? Please describe.**
Currently using CausalML package. 

I've difficulties understanding what is the impact of over-fitting on ATE. Dive in the package code to see that we are predicting on the same data we train. 
https://github.com/uber/causalml/pull/511

XGBoost is a popular regressor and widely showcased in the documentation, however it is known to easily overfit (because of boosting). 

What is the impact of over-fitting on ATE computation? I don't find any literaturee online about this (it is not related to CausalML 😉 ). 

When writing this issue, stepped on this PR  [Ate pretrain 0506](https://github.com/uber/causalml/pull/511 ) which is a first step for computing out-of-fold (or in cross-validation) the ATE.

**Describe the solution you'd like**
Documentation to point to relevant literature. 
Ability to `estimate_ate()` out-of-fold

**Describe alternatives you've considered**
I started implementing my own out-of-fold ATE with cross-validation on my side

```
        kf = KFold(n_splits=5, random_state=1, shuffle=True)
        te = []
        for train_index, valid_index in kf.split(final):
            # Can't used named is_train is get_x_y_treatment because different naming in different
            x_train, y_train, treatment_train = self.get_x_y_treatment(
                final.iloc[train_index], True
            )
            x_valid, y_valid, treatment_valid = self.get_x_y_treatment(
                final.iloc[valid_index], False
            )
            self.my_model.fit(x_train, treatment_train, y_train)
            te_valid = self.my_model.predict(x_valid, treatment_valid, y_valid)
            te.append(te_valid)

        te = np.concatenate(te).mean()
        lb, ub = np.nan, np.nan
```

The simplest would probably be to directly add this method to the package (to keep confidence interval and stuff) with something like this.

```
class BaseTLearner(...):

    def estimate_outoffold_ate(
        self,
        X_train,
        treatment_train,
        y_train,
        X_valid,
        treatment_valid,
        y_valid,
        p=None,
        bootstrap_ci=False,
        n_bootstraps=1000,
        bootstrap_size=10000,
    ):
        """Estimate the Average Treatment Effect (ATE).

        Args:
            X (np.matrix or np.array or pd.Dataframe): a feature matrix
            treatment (np.array or pd.Series): a treatment vector
            y (np.array or pd.Series): an outcome vector
            bootstrap_ci (bool): whether to return confidence intervals
            n_bootstraps (int): number of bootstrap iterations
            bootstrap_size (int): number of samples per bootstrap
        Returns:
            The mean and confidence interval (LB, UB) of the ATE estimate.
        """
        X_train, treatment_train, y_train = convert_pd_to_np(
            X_train, treatment_train, y_train
        )
        self.fit(X_train, treatment_train, y_train, return_components=True)

        X_valid, treatment_valid, y_valid = convert_pd_to_np(
            X_valid, treatment_valid, y_valid
        )
        te, yhat_cs, yhat_ts = self.predict(
            X_valid, treatment_valid, y_valid, return_components=True
        )

        ate = np.zeros(self.t_groups.shape[0])
        ate_lb = np.zeros(self.t_groups.shape[0])
        ate_ub = np.zeros(self.t_groups.shape[0])

        for i, group in enumerate(self.t_groups):
            _ate = te[:, i].mean()

            mask = (treatment_valid == group) | (treatment_valid == self.control_name)
            treatment_filt = treatment_valid[mask]
            y_filt = y_valid[mask]
            w = (treatment_filt == group).astype(int)
            prob_treatment = float(sum(w)) / w.shape[0]

            yhat_c = yhat_cs[group][mask]
            yhat_t = yhat_ts[group][mask]

            se = np.sqrt(
                (
                    (y_filt[w == 0] - yhat_c[w == 0]).var() / (1 - prob_treatment)
                    + (y_filt[w == 1] - yhat_t[w == 1]).var() / prob_treatment
                    + (yhat_t - yhat_c).var()
                )
                / y_filt.shape[0]
            )

            _ate_lb = _ate - se * norm.ppf(1 - self.ate_alpha / 2)
            _ate_ub = _ate + se * norm.ppf(1 - self.ate_alpha / 2)

            ate[i] = _ate
            ate_lb[i] = _ate_lb
            ate_ub[i] = _ate_ub

        if not bootstrap_ci:
            return ate, ate_lb, ate_ub
        else:
            t_groups_global = self.t_groups
            _classes_global = self._classes
            models_c_global = deepcopy(self.models_c)
            models_t_global = deepcopy(self.models_t)

            logger.info("Bootstrap Confidence Intervals for ATE")
            ate_bootstraps = np.zeros(shape=(self.t_groups.shape[0], n_bootstraps))

            for n in tqdm(range(n_bootstraps)):
                ate_b = self.bootstrap(
                    X_valid, treatment_valid, y_valid, size=bootstrap_size
                )
                ate_bootstraps[:, n] = ate_b.mean()

            ate_lower = np.percentile(
                ate_bootstraps, (self.ate_alpha / 2) * 100, axis=1
            )
            ate_upper = np.percentile(
                ate_bootstraps, (1 - self.ate_alpha / 2) * 100, axis=1
            )

            # set member variables back to global (currently last bootstrapped outcome)
            self.t_groups = t_groups_global
            self._classes = _classes_global
            self.models_c = deepcopy(models_c_global)
            self.models_t = deepcopy(models_t_global)

            return ate, ate_lower, ate_upper

```

Thoughts on this? Let me know if you think it is a good idea to implement this myself then open a PR 😅

**Comment by Jacques-Peeters** [comment]
Also implemented an XGBoost with default train/valid split and early stopping 
```
"""Custom models."""
from typing import Any

from sklearn.model_selection import train_test_split
from xgboost import XGBRegressor


class XGBRegressorEarlyStopping(XGBRegressor):
    """XGBoost with train/valid split and early stopping.

    Required wrapper to make it work seamlessly with CaulsalML.

    Args:
        XGBRegressor (_type_): XGBoost model.
    """

    def fit(
        self: "XGBRegressorEarlyStopping", x: Any, y: Any, *args: Any, **kwargs: Any
    ) -> None:  # noqa
        """Fit XGBRegressor with train/valid split and early stopping."""
        x_train, x_valid, y_train, y_valid = train_test_split(x, y, test_size=0.20)

        super().fit(
            x_train,
            y_train,
            eval_set=[(x_valid, y_valid)],
            verbose=False,
            *args,
            **kwargs
        )
```

Benchmarked on a proprietary dataset, with T-learner and two different categorical encoding (One-hot-encoding and out-of-fold target encoding) here are the results :
```
2022-06-08 10:25:44 INFO: Estimate ATE for: TargetEncoding(my_model=BaseTRegressor)
2022-06-08 10:26:44 INFO: Average Treatment Effect (out-of-fold + early_stopping): 0.20 (nan, nan)
2022-06-08 10:26:44 INFO: Estimate ATE for: TargetEncoding(my_model=BaseTRegressor)
2022-06-08 10:27:11 INFO: Average Treatment Effect (out-of-fold + early_stopping): 0.23 (nan, nan)
2022-06-08 10:27:11 INFO: Estimate ATE for: OneHotEncoding(my_model=BaseTRegressor)
2022-06-08 10:28:08 INFO: Average Treatment Effect (out-of-fold + early_stopping): 0.21 (nan, nan)
2022-06-08 10:28:08 INFO: Estimate ATE for: OneHotEncoding(my_model=BaseTRegressor)
2022-06-08 10:28:28 INFO: Average Treatment Effect (out-of-fold + early_stopping): 0.22 (nan, nan)
```

There are differences but are they significant?

**Comment by t-tte** [comment]
This is a great question. I'm sure that the effect of overfitting on CATE computation depends on the specific meta-learner that you are interested in. For example, the [X-learner paper](https://arxiv.org/pdf/1706.03461.pdf) discusses the issue briefly in the context of the T-learner.

In your issue, you're talking about CATE computation, but your code suggests that you're actually talking about ATE computation. Is that correct?

What is the main use case for the propose out-of-sample ATE prediction? Recall that the way in which models are currently evaluated is by making out-of-sample CATE predictions and using a metric such as AUUC to evaluate the models. This implicitly takes into account overfitting since we would expect overfitted meta-learners to yield poorer CATE predictions.

**Comment by t-tte** [comment]
Could you please give an example of the kind of use case that you're solving? "More accurate ATE" can mean a lot of different things. For example, are you dealing with experimental or observational data?

**Comment by aman-coder03** [comment]
hi @t-tte, stumbled upon this issue and wanted to revive it as I think it's still very relevant

to address the open question about use case...the need for out-of-fold ATE estimation applies to both experimental and observational settings. In A/B tests, overfitting in outcome models can inflate variance in CATE estimates and bias the ATE on finite samples. In observational settings the problem is worse since both propensity and outcome models are at risk

implementation proposed by @Jacques-Peeters looks solid as a starting point. I would be interested in picking this up and opening a PR if maintainers are still open to it. Happy to discuss scope and design before starting.

**Comment by jbbqqf** [comment]
Hi! Triaging older issues — I think this one is at least partially addressed and may be a candidate for closure.

The OOF-ATE work referenced in the issue body (#511 — "Ate pretrain 0506") landed and exposes the cross-validated path on most meta-learners (see `estimate_ate(..., pretrain=True)` and the tests under `tests/test_meta_learners.py::test_XGBTRegressor` for the OOF check). The broader theoretical question of overfitting impact on CATE/ATE is more a research-doc topic than a code change, and `causalml`'s docs already cite the relevant references for X-learner / DR-learner.

If there's a specific gap remaining (e.g. a particular learner where the OOF path is missing), a fresh issue with a focused ask would be easier to schedule than reopening this one.

(Disclosure: I drafted this comment with help from Claude Code while triaging stale issues; the references above were verified manually against the current `master` branch.)

**Comment by jeongyoonlee** [comment]
Current status on `master` (`0ba283e`): `estimate_ate(..., pretrain=True)` skips refitting and estimates from the already-fitted model, which supports the fit-on-train / estimate-on-validation workflow in the opening post.

Available on `BaseSLearner`, `BaseTLearner`, `BaseXLearner`, `BaseRLearner`, `BaseDRLearner`, `LRSRegressor`. Not available on `CausalTreeRegressor.estimate_ate` or `BaseDRIVLearner.estimate_ate`.

```python
learner.fit(X=X_train, treatment=w_train, y=y_train)
ate, lb, ub = learner.estimate_ate(X=X_valid, treatment=w_valid, y=y_valid, pretrain=True)
```

@aman-coder03 — if you still want to pick this up, the remaining scope is adding `pretrain` to those two classes. Worth a separate issue with the API sketch, since `BaseDRIVLearner` also carries `assignment` and `pZ`.

**Comment by jeongyoonlee** [comment]
Scope for the remaining work, against `a77dfb4`.

1. **`CausalTreeRegressor.estimate_ate`** — add `pretrain=False` as the last parameter. `pretrain=True` calls `self.predict(X)` in place of `self.fit_predict(X, treatment, y)`.
2. **`BaseDRIVLearner.estimate_ate`** — same, calling `self.predict(X, treatment, y, return_components=True)`. `assignment` and `pZ` are only used on the fitting path, so they go unused when `pretrain=True`.
3. `TMLELearner.estimate_ate` is out of scope: the class has no `fit`, so there is no fitted model to reuse. Those three are every `estimate_ate` in the package without `pretrain`; the other 22 classes have it.
4. `estimate_ate` carries the #854 argument-order shim on all three classes, so `pretrain` goes last in the signature.

Tests: one per class asserting that `pretrain=True` reuses the fitted model rather than refitting — same predictions before and after the call — and that the estimate differs from the in-sample one on held-out rows.

For context on what this changes: fitting on half the rows and estimating on the other half gave a mean absolute ATE error of 0.0990 against 0.1140 for the current in-sample path (20 seeds, `synthetic_data(mode=2, n=2000)`, `CausalTreeRegressor`).

Filed #1006 separately for the causal tree's confidence interval, which is wrong independently of this: `se = dhat.std() / X.shape[0]` divides by `n` rather than `sqrt(n)`, and covered the true ATE in 0 of 20 seeds. Worth landing before or alongside this, since `pretrain=True` otherwise moves an interval that does not cover.
